python/delta/tests/test_deltatable.py

```diff
@@ -850,6 +850,72 @@ def runRestoreToVersion() -> None:
         self.__intercept(runRestoreToVersion,
                          "version needs to be an int but got '<class 'str'>'")
 
+    def test_optimize(self) -> None:
+        # write an unoptimized delta table
+        df = self.spark.createDataFrame([("a", 1), ("a", 2)], ["key", "value"]).repartition(1)
```

Comment:
can we use the utility methods `__writeDeltaTable`, `__writeAsTable` or `__overwriteDeltaTable` here? Is the number of files hard to control without the repartition for test verification?

Reply:
@vkorukanti I checked these but I think I need something that writes unoptimized delta tables. I want to create a delta table with multiple parquet files, then I can execute compaction on them. I used ``repartition`` to control the number of files written, and I used ``partitionBy`` to test the partitioned table case.

I doubt ``__writeDeltaTable``, ``__writeAsTable`` or `` __overwriteDeltaTable`` helps when creating such setting.

Do you have any suggestions? I can remove "numFilesAdded" and "numFilesRemoved" verifications from the test if it is not necessary. I can use ``__writeDeltaTable`` that way.

Or I can add "mode", "repartition" and "partitioned" parameters to ``__writeDeltaTable``. So I can use it as:

```python
self.__writeDeltaTable([("a", 1), ("a", 2)], repartition=1)
self.__writeDeltaTable([("a", 3), ("a", 4)], repartition=1, mode="append")
self.__writeDeltaTable([("b", 1), ("b", 2)], repartition=1, mode="append")

# and
self.__writeDeltaTable([("a", 1), ("a", 2)], repartition=1, partitioned=True)
self.__writeDeltaTable([("a", 3), ("a", 4)], repartition=1, partitioned=True, mode="append")
self.__writeDeltaTable([("b", 1), ("b", 2)], repartition=1, partitioned=True, mode="append")

# where
def __writeDeltaTable(self, datalist, repartition=None, partitioned=False, mode="error"):
    df = self.spark.createDataFrame(datalist, ["key", "value"])
    if repartition:
        df = df.repartition(repartition)
    writer = df.write
    if partitioned:
        writer = writer.partitionBy("key")
    writer.format("delta").save(self.tempFile, mode=mode)
```

Let me what you think.

Reply:
Thanks for checking. I think the current method is fine. In future if there are more tests that need such utility methods, we can add them.